xen/sysctl: return per-node claimed memory in XEN_SYSCTL_numainfo

This allow the toosltack to calculate the real per-node available memory.
Missing the libxl changes to expose this information on the libxl API.

Signed-off-by: Roger Pau Monné <roger.pau@citrix.com>
diff --git a/xen/common/numa.c b/xen/common/numa.c
index 5df3a573ad68..fac8d3c9b728 100644
--- a/xen/common/numa.c
+++ b/xen/common/numa.c
@@ -702,10 +702,12 @@ static void cf_check dump_numa(unsigned char key)
     for_each_online_node ( i )
     {
         mfn_t mfn = _mfn(node_start_pfn(i) + 1);
+        unsigned long claimed, free;
 
-        printk("NODE%u start->%lu size->%lu free->%lu\n",
+        free = avail_node_heap_pages(i, &claimed);
+        printk("NODE%u start->%lu size->%lu free->%lu claimed->%lu\n",
                i, node_start_pfn(i), node_spanned_pages(i),
-               avail_node_heap_pages(i));
+               free, claimed);
         /* Sanity check mfn_to_nid() */
         if ( node_spanned_pages(i) > 1 && mfn_to_nid(mfn) != i )
             printk("mfn_to_nid(%"PRI_mfn") -> %d should be %u\n",
diff --git a/xen/common/page_alloc.c b/xen/common/page_alloc.c
index b84f11854d77..e61b243cd2c9 100644
--- a/xen/common/page_alloc.c
+++ b/xen/common/page_alloc.c
@@ -2038,7 +2038,8 @@ static void init_heap_pages(
 }
 
 static unsigned long avail_heap_pages(
-    unsigned int zone_lo, unsigned int zone_hi, unsigned int node)
+    unsigned int zone_lo, unsigned int zone_hi, unsigned int node,
+    unsigned long *claimed)
 {
     unsigned int i, zone;
     unsigned long free_pages = 0;
@@ -2055,6 +2056,9 @@ static unsigned long avail_heap_pages(
                 free_pages += avail[i][zone];
     }
 
+    if ( node != NUMA_NO_NODE && node < MAX_NUMNODES && claimed )
+        *claimed = per_node_outstanding_claims[node];
+
     return free_pages;
 }
 
@@ -2865,19 +2869,25 @@ unsigned long avail_domheap_pages_region(
     zone_hi = max_width ? bits_to_zone(max_width) : (NR_ZONES - 1);
     zone_hi = max_t(int, MEMZONE_XEN + 1, min_t(int, NR_ZONES - 1, zone_hi));
 
-    return avail_heap_pages(zone_lo, zone_hi, node);
+    return avail_heap_pages(zone_lo, zone_hi, node, NULL);
 }
 
 unsigned long avail_domheap_pages(void)
 {
     return avail_heap_pages(MEMZONE_XEN + 1,
                             NR_ZONES - 1,
-                            -1);
+                            -1, NULL);
 }
 
-unsigned long avail_node_heap_pages(unsigned int nodeid)
+unsigned long avail_node_heap_pages(unsigned int nodeid, unsigned long *claimed)
 {
-    return avail_heap_pages(MEMZONE_XEN, NR_ZONES -1, nodeid);
+    unsigned long avail;
+
+    spin_lock(&heap_lock);
+    avail = avail_heap_pages(MEMZONE_XEN, NR_ZONES -1, nodeid, claimed);
+    spin_unlock(&heap_lock);
+
+    return avail;
 }
 
 
@@ -2888,7 +2898,7 @@ static void cf_check pagealloc_info(unsigned char key)
 
     printk("Physical memory information:\n");
     printk("    Xen heap: %lukB free\n",
-           avail_heap_pages(zone, zone, -1) << (PAGE_SHIFT-10));
+           avail_heap_pages(zone, zone, -1, NULL) << (PAGE_SHIFT-10));
 
     while ( ++zone < NR_ZONES )
     {
@@ -2898,7 +2908,7 @@ static void cf_check pagealloc_info(unsigned char key)
             total = 0;
         }
 
-        if ( (n = avail_heap_pages(zone, zone, -1)) != 0 )
+        if ( (n = avail_heap_pages(zone, zone, -1, NULL)) != 0 )
         {
             total += n;
             printk("    heap[%02u]: %lukB free\n", zone, n << (PAGE_SHIFT-10));
diff --git a/xen/common/sysctl.c b/xen/common/sysctl.c
index c2d99ae12e84..4d577170655b 100644
--- a/xen/common/sysctl.c
+++ b/xen/common/sysctl.c
@@ -317,7 +317,7 @@ long do_sysctl(XEN_GUEST_HANDLE_PARAM(xen_sysctl_t) u_sysctl)
                     if ( node_online(i) )
                     {
                         meminfo.memsize = node_spanned_pages(i) << PAGE_SHIFT;
-                        meminfo.memfree = avail_node_heap_pages(i) << PAGE_SHIFT;
+                        meminfo.memfree = avail_node_heap_pages(i, NULL) << PAGE_SHIFT;
                     }
                     else
                         meminfo.memsize = meminfo.memfree = XEN_INVALID_MEM_SZ;
diff --git a/xen/include/xen/mm.h b/xen/include/xen/mm.h
index 4a0e493519c5..9f79ac076463 100644
--- a/xen/include/xen/mm.h
+++ b/xen/include/xen/mm.h
@@ -143,7 +143,8 @@ void free_domheap_pages(struct page_info *pg, unsigned int order);
 unsigned long avail_domheap_pages_region(
     unsigned int node, unsigned int min_width, unsigned int max_width);
 unsigned long avail_domheap_pages(void);
-unsigned long avail_node_heap_pages(unsigned int nodeid);
+unsigned long avail_node_heap_pages(unsigned int nodeid,
+                                    unsigned long *claimed);
 #define alloc_domheap_page(d,f) (alloc_domheap_pages(d,0,f))
 #define free_domheap_page(p)  (free_domheap_pages(p,0))
 unsigned int online_page(mfn_t mfn, uint32_t *status);
